src/main/java/dasniko/keycloak/authenticator/EmailAuthenticator.java

```diff
@@ -0,0 +1,73 @@
+package dasniko.keycloak.authenticator;
+
+import lombok.extern.slf4j.Slf4j;
+import org.keycloak.authentication.AuthenticationFlowContext;
+import org.keycloak.authentication.AuthenticationFlowError;
+import org.keycloak.email.EmailTemplateProvider;
+import org.keycloak.models.*;
+import org.keycloak.sessions.AuthenticationSessionModel;
+import org.keycloak.theme.Theme;
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+
+import static dasniko.keycloak.authenticator.OTPConstants.*;
+
+@Slf4j
+public class EmailAuthenticator extends OTPAuthenticator {
+	private static final String TPL_CODE = "login-email.ftl";
+	private static final String EMAIL_OTP_TEMPLATE = "email-otp.ftl";
+
+	@Override
+	public void authenticate(AuthenticationFlowContext context) {
+		AuthenticatorConfigModel config = context.getAuthenticatorConfig();
+		KeycloakSession session = context.getSession();
+		UserModel user = context.getUser();
+
+		int ttl = getTTL(config);
+		String code = getSecretCode(config);
+
+		AuthenticationSessionModel authSession = context.getAuthenticationSession();
+		authSession.setAuthNote(OTPConstants.CODE, code);
+		authSession.setAuthNote(OTPConstants.CODE_TTL, Long.toString(System.currentTimeMillis() + (ttl * 1000L)));
+
+		RealmModel realm = context.getRealm();
+
+		EmailTemplateProvider emailTemplateProvider = context.getSession().getProvider(EmailTemplateProvider.class);
+		emailTemplateProvider.setRealm(realm);
+		emailTemplateProvider.setUser(user);
+
+		try {
+			Theme theme = session.theme().getTheme(Theme.Type.LOGIN);
+			Locale locale = session.getContext().resolveLocale(user);
+			String otpAuthText = theme.getMessages(locale).getProperty("authCodeText");
+			String otpText = String.format(otpAuthText, code, Math.floorDiv(ttl, 60));
+
+			Map<String, Object> msgParams = new HashMap<>();
+			msgParams.put("code", otpText);
+
+			String subjectLine = theme.getMessages(locale).getProperty("emailAuthSubject");
+			emailTemplateProvider.send(subjectLine, EMAIL_OTP_TEMPLATE, msgParams);
```

Comment:
should we implement the `SIMULATION_MODE`? As described when set to true should print the message to server logs

Reply:
Actually i think i should remove that from the Factory configuration - for email, we can use mailhog for testing so there's no real need for it. It's set up in the SMS class so that we don't need to have a real provider setup and incur costs for initial testing. I'll remove from the email config and add a note - thanks for catching that!